### ISSUE TITLE
`jnp.unique` along an axis is very slow for large arrays

### ISSUE THREAD

**Issue by jakevdp** [google/jax#17370]
For example:
```
import jax.numpy as jnp
x = jnp.zeros((2, 1000))
jnp.unique(x, axis=0)  # takes ~2.5 minutes on a Colab CPU
```
The issue is that XLA has no `unique` primitive, so we're forced to implement it from scratch. `unique` along an axis is essentially implemened as a lexicographic sort followed by a pairwise equality check, and in this case the lex-sort splits the array columns into 1000 key arrays passed to XLA, and this has performance implications.

I brainstormed a bit with @mattjj about how we might better compute this in this case, but the answer isn't obvious.

I'm going to leave this open for now.

**Comment by jakevdp** [comment]
We might implement `unique` via spectral analysis; here's a rough example:
```python
def unique(x):
  x = jnp.asarray(x)
  d = x.shape[0]

  # Construct adjacency matrix
  A = (x[:, None] == x[None, :]).reshape(d, d, -1).all(-1)\

  # Construct Laplacian matrix & compute eigen-decomposition
  L = A - jnp.diag(A.sum(axis=1))
  evals, evecs = jnp.linalg.eigh(L)

  # Size of laplacian null space = number of unique values
  num_unique = (abs(evals) < 0.5).sum()

  # Nonzero entries in null-space eigenvectors => indices of unique values
  i = jnp.argmax(abs(evecs), axis=0)
  i = i[-num_unique:]  # Need to change this for JIT-compatibility

   # Note: return unsorted unique values because lexsort is what's causing issues!
  return x[i]
```
This works for 1D inputs:
```python
x = jnp.array([1, 2, 3, 2, 3, 2, 1])
print(unique(x))
# [3 2 1]
```
As well as N-dimensional inputs (`axis=0` is implied)
```python
x = jnp.zeros((100, 10)).at[0, 0].set(1)
print(unique(x))
# [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
#  [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
```

**Comment by MoFHeka** [comment]
Any progress? Is migrate the unique operator from TensorFlow as an independent primitive a good option? like TopK operator.

**Comment by jakevdp** [comment]
There hasn't been any progress on this unfortunately. As mentioned above, the `unique` implementation is based on XLA's variadic sort, and as the number of variables sorted becomes large, XLA slows down just as it does for any computation with many inputs & operations.

**Comment by jakevdp** [comment]
Without more information on what exactly you have in mind, it’s hard to evaluate whether it would be better.

**Comment by MoFHeka** [comment]
What about directly migrate hard codes from TensorFlow op kernel? Since the most common solution of Unique OP which is Radix sort and it’s hard to be expressed by XLA. 

https://github.com/tensorflow/tensorflow/blob/abcaba2e5fce335212d1f7617bdb4e3def525c23/tensorflow/core/kernels/unique_op_gpu.cu.h

**Comment by jakevdp** [comment]
The algorithm implemented in the tensorflow cuda kernel (described [here](https://github.com/tensorflow/tensorflow/blob/abcaba2e5fce335212d1f7617bdb4e3def525c23/tensorflow/core/kernels/unique_op_gpu.cu.h#L228)) is essentially identical to how it's implemented in XLA, so I'm not sure whether it would offer much room for improvement.

**Comment by MoFHeka** [comment]
https://github.com/c3sr/tcu_scope
what about a algorithm running reduce and scan kernel in Radix sort with Nvidia Tensor Core?

**Comment by jakevdp** [comment]
I'm not sure about the implications to GPU memory use – you'd probably have to profile it if you're interested in performance at that level.

**Comment by rosspeili** [comment]
Hey @jakevdp can i give it a spin also?
